src/Foundatio/Caching/InMemoryCacheClient.cs

```diff
@@ -847,23 +871,28 @@ private async Task StartMaintenanceAsync(bool compactImmediately = false)
         }
     }
 
-    private Task CompactAsync()
+    private async Task CompactAsync()
     {
         if (!_maxItems.HasValue || _memory.Count <= _maxItems)
-            return Task.CompletedTask;
+            return;
 
         string expiredKey = null;
-        lock (_lock)
+        using (await _lock.LockAsync().AnyContext())
         {
             if (_memory.Count <= _maxItems)
-                return Task.CompletedTask;
+                return;
 
             (string Key, long LastAccessTicks, long InstanceNumber) oldest = (null, Int64.MaxValue, 0);
             foreach (var kvp in _memory)
             {
-                if (kvp.Value.LastAccessTicks < oldest.LastAccessTicks
-                    || (kvp.Value.LastAccessTicks == oldest.LastAccessTicks && kvp.Value.InstanceNumber < oldest.InstanceNumber))
+                bool isExpired = kvp.Value.ExpiresAt < SystemClock.UtcNow;
```

Comment:
Optimized this to grab the first expired one (and break), this doesn't take into account the frequency period like maintenance loop.